### PR TITLE
Add HAL implementation hooks to cv::flip() and cv::rotate() functions from core module 

### PR DESCRIPTION
**Pull request by jvuillaumier**
Hello,

This change proposes the addition of HAL hooks for cv::flip() and cv::rotate() functions from OpenCV core module.
Flip and rotation are functions commonly available from 2D hardware accelerators. This is convenient provision to enable custom optimized implementation of image flip/rotation on systems embedding such accelerator.

Thank you
 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@vpisarev @opencv-alalek could you take a look?

**Comment by jvuillaumier** [comment]
Change seems to break videoio/VideoCaptureAPITests.mp4_orientation_meta_auto test.
My bad - I am looking at it and will update.

**Comment by asmorkalov** [comment]
videoio/VideoCaptureAPITests.mp4_orientation_meta_auto uses `cv::rotate` internally.

**Comment by jvuillaumier** [comment]
`cv::rotate` is used internally indeed, when a rotation meta  is present in the stream:
https://github.com/opencv/opencv/blob/4.x/modules/videoio/src/cap_interface.hpp#L269

In place rotation is used, that may be a case where this change causes problem.
I will check what the actual issue is and update.

**Comment by jvuillaumier** [comment]
Hello,

Change has been updated to take care of the in-place `cv::rotate()` case where same Mat is used for source and destination.

Test `videoio/VideoCaptureAPITests.mp4_orientation_meta_auto` is fixed.

Thanks

**Comment by jvuillaumier** [comment]
Previous failure on `videoio/VideoCaptureAPITests.mp4_orientation_meta_auto` is confirmed to be fixed (in place case).

However CI suggests an abort was detected running `stitchExposureCompMultiFeed_a123.a123/12` test on macOS-x64  - this test is passed on other platforms.

I will look at it and update.

**Comment by jvuillaumier** [comment]
Hello,

I had a closer look to the abort issue reported by CI on `stitchExposureCompMultiFeed_a123.a123/12` performance test on macOS-x64.

It is a bit puzzling me because:
- Neither this test nor any other test from `opencv_perf_stitching` test suite is calling directly or indirectly a function changed by this patch (flip/rotate)
- Test is passed on other platforms from CI
- I could not repro the error locally neither on Yocto-ARM64 nor on Ubuntu-x64
- Looking at recent PRs, I have not spotted such error so this test is presumably deemed stable

As I do not have access to a macOS-x64 platform, is it possible to restart CI on same change to confirm the issue and rule out a CI glitch?

Thanks

**Review by asmorkalov — APPROVED** [review]
:+1:

File: modules/core/include/opencv2/core/hal/hal.hpp [hunk 1]
```diff
@@ -201,6 +201,9 @@ CV_EXPORTS void cvt32f16f( const float* src, float16_t* dst, int len );
 CV_EXPORTS void addRNGBias32f( float* arr, const float* scaleBiasPairs, int len );
 CV_EXPORTS void addRNGBias64f( double* arr, const double* scaleBiasPairs, int len );
 
+CV_EXPORTS void flip( int src_type, const uchar* src_data, size_t src_step, int src_width, int src_height, uchar* dst_data, size_t dst_step, int flip_mode );
+CV_EXPORTS void rotate( int src_type, const uchar* src_data, size_t src_step, int src_width, int src_height, uchar* dst_data, size_t dst_step, int angle );
```
Comment (opencv-alalek) [inline modules/core/include/opencv2/core/hal/hal.hpp:205]: All functions from this file are defined somewhere in OpenCV codebase.
Added function are not implemented. So they should not be here.

File: modules/core/src/matrix_transform.cpp [hunk 2]
```diff
@@ -1097,4 +1101,50 @@ void rotate(InputArray _src, OutputArray _dst, int rotateMode)
     }
 }
 
+void rotate(InputArray _src, OutputArray _dst, int rotateMode)
+{
+    int angle;
```
Comment (opencv-alalek) [inline modules/core/src/matrix_transform.cpp:1106]: Missing checks of input parameters constraints.

File: modules/core/src/hal_replacement.hpp [hunk 3]
```diff
@@ -731,6 +731,47 @@ inline int hal_ni_minMaxIdx(const uchar* src_data, size_t src_step, int width, i
...
+   If angle has value [180] it is also destination image width
+   If angle has values [90, 270] it is also destination image height
+   @param src_height source and destination image height (destination image width for angles [90, 270])
+   If angle has value [180] it is also destination image height
+   If angle has values [90, 270] it is also destination image width
+   @param dst_data destination image data
+   @param dst_step destination image step
+   @param angle clockwise angle for rotation in degrees from set [90, 180, 270]
+ */
+inline int hal_ni_rotate(int src_type, const uchar* src_data, size_t src_step, int src_width, int src_height,
```
Comment (opencv-alalek) [inline modules/core/src/hal_replacement.hpp:768]: `rotate` => `rotate90`

To avoid confusion with generic rotation.

**Review by vpisarev — APPROVED** [review]

File: modules/core/include/opencv2/core/hal/hal.hpp (hunk 1 above)
Comment (jvuillaumier) [reply modules/core/include/opencv2/core/hal/hal.hpp]: Lines removed in updated version of the patch
Thanks

File: modules/core/src/matrix_transform.cpp (hunk 2 above)
Comment (jvuillaumier) [reply modules/core/src/matrix_transform.cpp:1105]: Check present in initial version of cv::rotate() is now present in the updated patch.
Thanks

File: modules/core/src/hal_replacement.hpp (hunk 3 above)
Comment (jvuillaumier) [reply modules/core/src/hal_replacement.hpp]: Functions renamed accordingly in updated version of the patch:
`hal_ni_rotate` => `hal_ni_rotate90`
`cv_hal_rotate` => `cv_hal_rotate90`

Thanks

**Comment by jvuillaumier** [comment]
CI reports test failures in DNN module for multiple platforms.
Updates in last patch version consist in cleanup, function/macro renaming and additional input parameter check in  `cv::rotate()` - mostly non functional changes.

As CI status for the previous version of the patch was green, I presume present CI test failures are not related to this patch version.

**Comment by jvuillaumier** [comment]
CI errors for the DNN Einsum tests are likely related to #24311

**Review by opencv-alalek — APPROVED** [review]

**Pull request merged**